### ISSUE TITLE
Tasks disappear in OpenTasks app when closed manually

### ISSUE THREAD

**Issue by akn01** [dmfs/opentasks#404]
Hello,

I'm using DAVdroid 1.6.5-play on my Samsung Galaxy A3 (2017) with Android 6.0.1 to synchronize Calendar, Contacts and Tasks with nextcloud 12.0.1.

Every time I close the OpenTask app manually and open it again, all tasks are gone. This also happes again and again by itself after some minutes. Then DAVdroid starts to synchronize them again.

I already reinstalled the OpenTasks app and DAVdroid.

Another phenomenon today: All tasks twice in OpenTasks app. They all disappear again when I close it. Then synchronization as usual...

**Comment by dmfs** [comment]
which OpenTasks version do you use?

**Comment by akn01** [comment]
I'm using version 1.1.9.

**Comment by Irihapeti** [comment]
I'm seeing this issue in Android 6.0.1.  In addition, all my tasks disappear as soon as I go turn off wifi or go out of range.

I have another phone running CM10.2 (Android 4.3.1), where I'm not seeing this problem at all. Both phones are using version 1.1.9.

**Comment by dmfs** [comment]
I'll add a couple of log statements which should help to identify what triggers the removal.

**Comment by dmfs** [comment]
Do you happen to know how to build the app yourself?

**Comment by dmfs** [comment]
One more question, is it the tasks which are gone (leaving empty task lists behind) or are the task lists gone as well?

**Comment by Irihapeti** [comment]
Sorry, I don't know how to build Android apps.

In answer to your second question, the task lists disappear entirely. If I try to add a task, I get "no task list available".

**Comment by akn01** [comment]
Same here.

**Comment by Trogel** [comment]
Same here on LineageOS 14.1-2017-08-09 (Android 7.1.2) with OpenTasks 1.1.9 and DavDroid 1.6.5-ose from F-Droid. I also built OpenTasks from the repository head and got the same issue.

Then I added debug logging into org.dmfs.provider.tasks.Utils.cleanUpLists(). It shows that this method gets called with an empty 'accounts' array. Strange! Does the system temporarily think there were no task accounts? In any case OpenTasks consequently deletes all entries from the 'Lists' table, etc.

To reproduce this issue, I wait until OpenTasks has completed syncing the tasks (coming from 6 lists on 2 accounts on an ownClound server). Then I add a fresh OpenTasks widget to the home screen that shows all tasks. Then I close the OpenTasks app by wiping it off the recent apps list. After some time or as soon as I touch the title bar of the widget, all tasks disappear and I get the log messages mentioned above. After few seconds, all tasks are pulled in again.

In addition, I added an aCalendar+ widget to the home screen. I noticed that it also loses all tasks at the same instant as OpenTasks, but the calendar entries remain intact.

@dmfs, please let me know, if I can help with further tests.

**Comment by dmfs** [comment]
"Empty account array" sounds like a permission issue. If the app doesn't have the `GET_ACCOUNTS` permission, it won't see the accounts managed by other apps. Instead it will just receive its own accounts (which is of course: none), hence all synced task lists are considered to be orphaned and deleted.
Please check if you've revoked the `GET_ACCOUNTS` permission. Due to a misconception in Android this permission is in the "Contacts" permission group (see https://issuetracker.google.com/issues/37065963) so you actually have to grant the "Contacts" permission.
I guess we should check if we have the `GET_ACCOUNTS` permission before we clean the lists up and show a message otherwise.

**Comment by Trogel** [comment]
Yes, that's it. In the permissions settings the option "Contacts" (or "Kontakte" in my case) was off. Turning it on fixed this issue! No more lost tasks!

I agree that it would help to get a message in case of missing permissions.

Many thanks for your quick and helpful response!

**Comment by Irihapeti** [comment]
Turning on contacts in permissions has fixed it here also. I notice that the older (CM) phone already had the contacts permissions turned on, which might account for the fact that it was working correctly.

Thanks also for the quick response.

**Comment by dmfs** [comment]
@lemonboston the app should check for the `GET_ACCOUNTS` permission on every start. If the permission has been granted it should continue, otherwise it should show a message about this issue and a "Grant" button which then launches the permission request. If the permission has been denied with "don't ask again" the message should instruct the user to go to settings and grant the permission.

Also check if we're using other runtime permissions.

**Comment by akn01** [comment]
> Yes, that's it. In the permissions settings the option "Contacts" (or "Kontakte" in my case) was off. Turning it on fixed this issue! No more lost tasks!

Right, problem solved. Thank you!

**Comment by lisetauber** [comment]
I use a Lenovo P2, OpenTasks 1.1.9, and I synchronise tasks with OpenSync . If the Internet connexion is off, all tasks disappear from OpenTasks. I have to redo a new synchronisation to recover the tasks. 

After a moment, with or without Internet connexion, all tasks disappear from the widget.

**Comment by smithfred** [comment]
dmfs, can you summarise https://issuetracker.google.com/issues/37065963 ? It's walled off behind a Google sign-in page (presumably because "reasons" :J)

**Comment by dmfs** [comment]
here is a copy of the original report: 
>Reported Issue
>*Affected versions:* 6.0.0
>*Affected devices:* most likely all, tested on Nexus 6
>
>*To reproduce:*
>
>1. Create project targeting SDK 23
>2. Declare using GET_ACCOUNTS permission in Manifest:
>
>    <uses-permission android:name="android.permission.GET_ACCOUNTS" />
>
>3. Ask for permission in runtime:
>
>    if (ContextCompat.checkSelfPermission(MainActivity.this,
>            Manifest.permission.GET_ACCOUNTS) != PackageManager.PERMISSION_GRANTED) {
>        if (ActivityCompat.shouldShowRequestPermissionRationale(MainActivity.this, Manifest.permission.GET_ACCOUNTS)) {
>            // TODO
>        } else {
>            ActivityCompat.requestPermissions(MainActivity.this, new String[] {Manifest.permission.GET_ACCOUNTS}, 23); // my lucky number
>        }
>    }
> 
>4. Run the app, observe "Allow <app name> to access your contacts?" dialog message which is totally unrelated to the permission granted to the app. After allowing this permission, app can list accounts of all types by calling AccountManager.get(context).getAccounts().
>
>This is because of GET_ACCOUNTS permission being member of CONTACTS group what seems a little bit awkward to me. Most likely GET_ACCOUNTS permission should be a member of other permission group. It's also a security/privacy issue, users are not aware they are granting the permission for reading their accounts.

And this was the response from Google:
>Status: Won't Fix (Intended Behavior)
>
>There is no plan to break out GET_ACCOUNTS in its own component in the Android M. However, this may be considered for future releases.

The issue has been reported on Oct 13th 2015 and marked as "fixed" on Oct 19th 2015. We can celebrate its 2nd anniversary soon.

**Comment by smithfred** [comment]
Haha, that's a nice total failure to acknowledge the legitimate issue there by Google 🤣

Thanks for the summary!

**Comment by dazzza** [comment]
Oreo has the contacts permission split.  On my Nexus 6P running AOSP its possible to grant Contacts permission, then click on it, then toggle off 'read your contacts' allowing access only to 'find accounts on device', interestingly the accounts permission doesn't have its own toggle

**Comment by dazzza** [comment]
To avoid the app getting access to read my contacts I forced stopped it before toggling the contacts toggle on and the read your contacts toggle off

**Comment by dmfs** [comment]
so `GET_ACCOUNTS` is still in the `Contacts` group but you can grant/revoke individual permissions of a group?

**Comment by dazzza** [comment]
Yeah that's right. Only other permission that's split appears to be phone, which now has a number of sub-settings. Contacts just has the two mentioned plus 'write to contacts'

**Comment by Offerel** [comment]
Ahhh now I know why my task list was empty after I installed manually 1.1.9....

I had installed first the version from Google Play store, where the latest version seems to be at 1.1.8.2. with this version contact permission was given on my device (Galaxy S7 with Android 7.0).

After I have seen that there is a newer version not available on Google play, I have uninstalled the Play store version and manually installed 1.1.9. Here the contact permission was not automatically given. Also the App does not ask for this permission.
